src/Common/src/System/Dynamic/Utils/CacheDict.cs

```diff
@@ -11,42 +12,71 @@ namespace System.Dynamic.Utils
     /// <summary>
     /// Provides a dictionary-like object used for caches which holds onto a maximum
     /// number of elements specified at construction time.
-    /// 
-    /// This class is not thread safe.
     /// </summary>
     internal class CacheDict<TKey, TValue>
     {
-        private readonly Dictionary<TKey, KeyInfo> _dict = new Dictionary<TKey, KeyInfo>();
-        private readonly LinkedList<TKey> _list = new LinkedList<TKey>();
-        private readonly int _maxSize;
+
+        // cache size is always ^2. 
+        // items are placed at [hash ^ mask]
+        // new item will displace previous one at the same location.
+        private readonly int mask;
+        private readonly Entry[] entries;
```

Comment:
I understand from your PR comments that you need this code to match what's in an internal branch.  Will we separately re-update this to fix up the names to match the style guidelines, e.g. starting these fields with underscores?